test/hotspot/jtreg/gc/TestReferenceRefersTo.java

```diff
@@ -243,10 +236,9 @@ private static void testConcurrentCollection() throws Exception {
                 fail("testWeak2 not notified");
             } else if (testWeak3 == null) {
                 fail("testWeak3 notified");
-            } else if (testWeak4 == null) {
-                if (obj4 != null) {
-                    fail("testWeak4 notified");
-                }
+            }
+            if ((testWeak4 == null) != (obj4 == null)) {
+                fail("either referent is cleared and we got notified, or neither of this happened");
```

Comment:
It might be helpful if the failure reported which one was non-null.  Not that it should ever fail...